.github/workflows/e2e.yml

```diff
@@ -118,12 +118,26 @@ jobs:
 
     steps:
       - uses: actions/checkout@v4
-      - name: Bootstrap
-        uses: ./.github/actions/bootstrap
+      - name: Set up Python
+        uses: actions/setup-python@v5
         with:
           python-version: 3.8
+      - name: Install build tools
+        run: |
+            python -m pip install -U pip==23.3.1
+        shell: bash
+      # Using approach described here for Python location caching:
+      # https://blog.allenai.org/python-caching-in-github-actions-e9452698e98d
+      - name: Cache Python location
+        id: cache-python
+        uses: actions/cache@v4
+        with:
+          path: ${{ env.pythonLocation }}
+          key: pythonloc-${{ runner.os }}-${{ matrix.directory }}-${{ env.pythonLocation }}-${{ hashFiles('**/pyproject.toml') }}
+          restore-keys: |
+            pythonloc-${{ runner.os }}-${{ matrix.directory }}-${{ env.pythonLocation }}
       - name: Install dependencies
-        run: python -m poetry install
+        run: python -m pip install --upgrade .
```

Comment:
Is there a reason why we do an upgrade while having [exact versions](https://github.com/adap/flower/pull/3480/files#diff-676c487c6f2c63c5d04f661140ceecdc515a4f04e074ec6415cb26b52439b1e1R12) ?

Reply:
Yes, in some `pyproject.toml`, we provide an upper limit to the versions allowed, such as `"torchvision>=0.14.1,<0.15.0"`. When the cache goes stale, `--upgrade` will install the latest allowed versions. At least this is my understanding from [this blogpost](https://blog.allenai.org/python-caching-in-github-actions-e9452698e98d).